ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -310,100 +311,152 @@ forkBlockProduction
-
-          newBlock <- atomically $ runProtocol varDRG $
-            produceBlock
-              proof
-              extLedger
-              currentSlot
-              prevPoint
-              prevNo
-              txs
-
-          trace $ TraceForgeEvent currentSlot newBlock
-          -- Adding a block is synchronous
-          ChainDB.addBlock chainDB newBlock
-          -- Check whether we adopted our block
-          curTip <- atomically $ ChainDB.getTipPoint chainDB
-          if curTip == blockPoint newBlock then do
-            trace $ TraceAdoptedBlock currentSlot newBlock txs
-          else do
-            isInvalid <- atomically $
-              ($ blockHash newBlock) . forgetFingerprint <$>
-              ChainDB.getIsInvalidBlock chainDB
-            case isInvalid of
-              Nothing ->
-                trace $ TraceDidntAdoptBlock currentSlot newBlock
-              Just reason -> do
-                trace $ TraceForgedInvalidBlock currentSlot newBlock reason
-                -- We just produced a block that is invalid according to the
-                -- ledger in the ChainDB, while the mempool said it is valid.
-                -- There is an inconsistency between the two!
-                --
-                -- Remove all the transactions in that block, otherwise we'll
-                -- run the risk of forging the same invalid block again. This
-                -- means that we'll throw away some good transactions in the
-                -- process.
-                removeTxs mempool (map txId txs)
+      withEarlyExit_ $ go currentSlot varDRG
   where
-    trace :: TraceForgeEvent blk (GenTx blk) -> m ()
-    trace = traceWith (forgeTracer tracers)
+    go :: SlotNo -> StrictTVar m PRNG -> WithEarlyExit m ()
+    go currentSlot varDRG = do
+        trace $ TraceStartLeadershipCheck currentSlot
+
+        -- Figure out which block to connect to
+        --
+        -- Normally this will be the current block at the tip, but it may
+        -- be the /previous/ block, if there were multiple slot leaders
+        (prevPoint, prevNo) <- do
+          mPrev <- lift $ atomically $ prevPointAndBlockNo currentSlot <$>
+                     ChainDB.getCurrentChain chainDB
+          case mPrev of
+            Right prev       -> return prev
+            Left  futureSlot -> do
+              trace $ TraceBlockFromFuture currentSlot futureSlot
+              exitEarly
+
+        -- Get ledger state corresponding to prevPoint
+        --
+        -- Under normal circumstances this should not fail. If we are unable to
+        -- get this ledger state, it means we switch to another chain which is
+        -- so distant or so much longer that the point we just looked up isn't
+        -- within @k@ blocks from the tip anymore. It's a theoretical possibility,
+        -- but very unlikely. When it happens, we simply give up on the chance
+        -- to produce a block.
```

Comment:
switch/switched

Isn't there the rather straightforward possibility of switching to a different short fork that does not include `prevPoint`? For example roll back 1, roll forward 2, between calling `ChainDB.getCurrentChain` and `ChainDB.getPastLedger`

Reply:
Ah, yes, you are right. I will update the comment.